### ISSUE TITLE
Reflection fails on .NET Native, resulting in runtime exception

### ISSUE THREAD

**Issue by atconrad22** [TalAloni/StandardSocketsHttpHandler#20]
When running on .NET Native (In my case, a UWP Release build with the .NET Native toolchain enabled) reflection on the core .net binaries is not allowed, so the following code always results in headersField being null, this ends up throwing an exception at runtime when it is used.

https://github.com/TalAloni/StandardSocketsHttpHandler/blob/d5233ccbe465621e35b594c40aa1c6b9df66f28e/StandardSocketsHttpHandler/Net/Http/Extensions/HttpRequestMessageExtensions.cs#L14-L20

As far as I can tell, the HasHeaders method is used as an optimization to prevent allocation of the the header objects when just trying to check if they exist, so it seems like it would be safe to take the small perf hit when running on .NET native to enable use of this library.

**Comment by TalAloni** [comment]
Hi,
"As far as I can tell, the HasHeaders method is used as an optimization to prevent allocation of the the header objects when just trying to check if they exist" -> I'm not sure I agree with that.

It seems to be the solution would be for you to fork the repository and create a compatible version to your platform, are you seeing any generic solution? what are you proposing?

**Comment by atconrad22** [comment]
Hi! Great library btw, really fixed some issues for me.

> "As far as I can tell, the HasHeaders method is used as an optimization to prevent allocation of the the header objects when just trying to check if they exist" -> I'm not sure I agree with that.

Oh really? I must be missing something else it is being used for then. Mind explaining what that is?

> It seems to be the solution would be for you to fork the repository and create a compatible version to your platform, are you seeing any generic solution? what are you proposing?

Well, if it was just for optimization, then adding a check like the one you already do for .NET Framework
https://github.com/TalAloni/StandardSocketsHttpHandler/blob/d5233ccbe465621e35b594c40aa1c6b9df66f28e/StandardSocketsHttpHandler/Net/Http/Extensions/HttpRequestMessageExtensions.cs#L12
Something like IsDotNetNative() and if it is, just return true from HasHeaders(). I made a fork doing this to get unblocked and it is working great for me so far, but I wanted to file an actual issue on it in case you are interested in me creating a PR instead of forking. But if you say HasHeaders() is more important than I thought then there could be underlying issues I am not seeing.

**Comment by TalAloni** [comment]
Thanks for the answer, how do you implement IsDotNetNative() ?

**Comment by atconrad22** [comment]
I just did nearly the exact same thing you did for IsDotNetFramework()

```
public static bool IsDotNetNative()
{
    const string DotnetNativeDescription = ".NET Native";
    string frameworkDescription = RuntimeInformation.FrameworkDescription;
    return frameworkDescription.StartsWith(DotnetNativeDescription);
}
```

**Comment by TalAloni** [comment]
Thanks,
and you didn't notice any issue when you always return true in such case?
and you don't have any other issue running under .NET Native?

**Comment by atconrad22** [comment]
I haven't seen any issues yet. Like I mentioned, it seems like HasHeaders() is primarily used to short circuit if statements to keep from allocating the Headers if they don't already exist. If so, then it's probably pretty safe to do this. Is there another use for HasHeaders() I missed? If so, then I might need to be a bit more cautious.

Haven't seen any other .NET Native related issues either. The only thing that made me need to make this change in the first place is that ".NET Native doesn't allow you to reflect over private members of the .NET Framework class library." as stated here: https://learn.microsoft.com/en-us/windows/uwp/dotnet-native/reflection-and-net-native

**Comment by TalAloni** [comment]
Thanks.
Which exception is thrown at runtime under .NET Native by the current code? (perhaps catching that exception is a better solution)

**Comment by atconrad22** [comment]
The reflection itself doesn't throw anything, it just returns null on the assignments to headersField, so it ends up throwing a NullReferenceException on the headersField.GetValue call at the end of HasHeaders. Can't say I'm super wild about throwing an exception on every single call to HasHeaders since that will end up being a bunch for every http request because the function is called multiple times.

**Comment by TalAloni** [comment]
Hi,
Please test my latest changes and report back on whether this address the issue.

Thanks!

**Comment by atconrad22** [comment]
I will give it a try as soon as I get a chance and update you with the results. Thanks for taking the time to look into this.

**Comment by TalAloni** [comment]
I've published v2.2.0.5 which should address this issue. Hopefully this issue is now resolved.
